pkg/querier/querier.go

```diff
@@ -133,52 +126,106 @@ func (q *Querier) SelectSamples(ctx context.Context, params logql.SelectSamplePa
 		newParams := logql.SelectSampleParams{
 			SampleQueryRequest: &queryRequestCopy,
 		}
-		newParams.End = adjustedEnd
-		chunkStoreIter, err = q.store.SelectSamples(ctx, newParams)
+		newParams.Start = ingesterQueryInterval.start
+		newParams.End = ingesterQueryInterval.end
+
+		ingesterIters, err := q.ingesterQuerier.SelectSample(ctx, newParams)
 		if err != nil {
 			return nil, err
 		}
-	default:
-		chunkStoreIter = iter.NoopIterator
 
-	}
-	// skip ingester queries only when QueryIngestersWithin is enabled (not the zero value) and
-	// the end of the query is earlier than the lookback
-	if !shouldQueryIngester(q.cfg, params) {
-		return chunkStoreIter, nil
+		iters = append(iters, ingesterIters...)
 	}
 
-	iters, err := q.ingesterQuerier.SelectSample(ctx, params)
-	if err != nil {
-		return nil, err
+	if storeQueryInterval != nil {
+		params.Start = storeQueryInterval.start
+		params.End = storeQueryInterval.end
+
+		storeIter, err := q.store.SelectSamples(ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		iters = append(iters, storeIter)
 	}
-	return iter.NewHeapSampleIterator(ctx, append(iters, chunkStoreIter)), nil
+	return iter.NewHeapSampleIterator(ctx, iters), nil
 }
 
-func shouldQueryIngester(cfg Config, params logql.QueryParams) bool {
-	lookback := time.Now().Add(-cfg.QueryIngestersWithin)
-	return !(cfg.QueryIngestersWithin != 0 && params.GetEnd().Before(lookback))
+func (q *Querier) buildQueryIntervals(queryStart, queryEnd time.Time) (*interval, *interval) {
+	// limitQueryInterval is a flag for whether store queries should be limited to start time of ingester queries.
+	limitQueryInterval := false
+	// ingesterMLB having -1 means query ingester for whole duration.
+	ingesterMLB := time.Duration(-1)
+	if q.cfg.IngesterQueryStoreMaxLookback != 0 {
+		// IngesterQueryStoreMaxLookback takes the precedence over QueryIngestersWithin while also limiting the store query range.
+		limitQueryInterval = true
+		ingesterMLB = q.cfg.IngesterQueryStoreMaxLookback
+	} else if q.cfg.QueryIngestersWithin != 0 {
+		ingesterMLB = q.cfg.QueryIngestersWithin
+	}
+
+	// query ingester for whole duration.
+	if ingesterMLB == -1 {
+		i := &interval{
+			start: queryStart,
+			end:   queryEnd,
+		}
+
+		if limitQueryInterval {
+			// query only ingesters.
+			return i, nil
+		}
+
+		// query both stores and ingesters without limiting the query interval.
+		return i, i
+	}
+
+	// see if there is an overlap between ingester query interval and actual query interval, if not just do the store query.
+	ingesterOldestStartTime := time.Now().Add(-ingesterMLB)
+	if queryEnd.Before(ingesterOldestStartTime) {
+		return nil, &interval{
+			start: queryStart,
+			end:   queryEnd,
+		}
+	}
+
+	// if there is an overlap and we are not limiting the query interval then do both store and ingester query for whole query interval.
+	if !limitQueryInterval {
+		i := &interval{
+			start: queryStart,
+			end:   queryEnd,
+		}
+		return i, i
+	}
```

Comment:
The microservices deployment mode with object store will run here no matter if there is an overlap.

Reply:
Use default configs:  ingester_config.query_store_max_look_back_period is 0

Reply:
This case will alse query ingester for whole duration.